File: solutions/erlang/gigasecond/1/gigasecond.erl
-module(gigasecond).
-export([from/1]).

-spec from(calendar:date() | calendar:datetime()) -> calendar:datetime().
from({Year, Month, Day}) ->
  from({{Year, Month, Day}, {0, 0, 0}});
from(DateTime) ->
  Gs = calendar:datetime_to_gregorian_seconds(DateTime) + 1000000000,
  calendar:gregorian_seconds_to_datetime(Gs).

File: solutions/elixir/gigasecond/1/lib/gigasecond.ex
defmodule Gigasecond do
  @doc """
  Calculate a date one billion seconds after an input date.
  """
  @spec from({{pos_integer, pos_integer, pos_integer}, {pos_integer, pos_integer, pos_integer}}) ::
          :calendar.datetime()

  def from(start_date = {{year, month, day}, {hours, minutes, seconds}}) do
    :calendar.datetime_to_gregorian_seconds(start_date) + 1_000_000_000
    |> :calendar.gregorian_seconds_to_datetime
  end
end
